### ISSUE TITLE
Unable to signin with Conditional Access flow

### ISSUE THREAD

**Issue by bgore** [msintuneappsdk/ms-intune-app-sdk-android#224]
# Intune Android App SDK Policy Enforcement Issue

## Summary

I have a similar problem to [Issue #87]: https://github.com/msintuneappsdk/ms-intune-app-sdk-android/issues/87, but not only do I not get the ```MsalIntuneAppProtectionPolicyRequiredException```, I do not complete enrollment or receive an access token.

The test org is internal, so I have access, but am unsure how to check everything for compliance.  In theory, the site has been setup according to:
Intune app protection polices have been applied to the app: https://learn.microsoft.com/en-us/mem/intune/apps/app-protection-policies
The CA policy was created: https://learn.microsoft.com/en-us/mem/intune/protect/app-based-conditional-access-intune-create

The auth_config.json file has: ```
{ "client_capabilities": "protapp", "client_id": private, "authorization_user_agent": "DEFAULT", "account_mode": "SINGLE", "redirect_uri": private, "broker_redirect_uri_registered": true, "authorities": [{ "type": "AAD", "audience": { "type": "AzureADMultipleOrgs", "tenant_id": "organizations" }}]}```
and ```AcquireSilentTokenParameters.Builder()``` includes ```.forceRefresh(false)```

## Repro Steps

1. Android emulator with client application and MS Company Portal installed
2. Client app is launched
3. Intune/Client email is gathered and enrollment started via:
```
        msalClientApplication?.let {
            val scopes = listOf("User.Read")
            val singleClientApplication = msalClientApplication as ISingleAccountPublicClientApplication
            val params = SignInParameters.builder()
                .withScopes(scopes)
                .withCallback(AuthCallback())
                .withLoginHint(email)
                .withActivity(fromActivity)
                .build()
            singleClientApplication.signIn(params)
```
4. The user password is entered in a WebView

**Expectation:**

Expected ```AuthenticationCallback.onError()``` to be called with a ```MsalIntuneAppProtectionPolicyRequiredException```

**Observation:**

Instead, receive a 'You can't get there from here' WebView.  Clicking OK causes ```AuthenticationCallback.onCancel()``` to be called.

![Screenshot from 2024-03-26 13-02-58](https://github.com/msintuneappsdk/ms-intune-app-sdk-android/assets/58744202/ccf9a4e9-7429-4f7a-bd5f-19eb70b2252e)

## Details

- **Intune Android App SDK Version:** - 10.0.0
- **Android Device Make and Model:** - Android Studio emulator
- **Android Device OS Version:** - x86, Android 11, API 30
- **Android Studio Version:** - Android Studio Giraffe | 2022.3.1 Patch 4

### Company Portal Logs

**Incident ID:**

5FJKX6T9

**Comment by mukeshk-ms** [comment]
@bgore  Based on the logs, it seems you should be getting the exception MsalIntuneAppProtectionPolicyRequiredException. Can you please check if you have handled it as per the [doc ](https://learn.microsoft.com/en-us/mem/intune/developer/app-sdk-android-phase7#handle-non-compliance-with-msal) and the sample [here ](https://github.com/msintuneappsdk/Taskr-Sample-Intune-Android-App/blob/c11ee86771f7dbc537b2e19e88eaed77ae06b4f0/app/src/main/java/com/microsoft/intune/samples/taskr/MainActivity.java#L206)

**Comment by bgore** [comment]
Yes, I have read that documentation and implemented the check for MsalIntuneAppProtectionPolicyRequiredException in the onError(MsalException) callback.  But as note, onError() is not called--the application blocks on the screen above.  Clicking OK causes onCancel() to be called.

**Comment by mukeshk-ms** [comment]
@bgore can you please confirm if you followed the MSAL prerequisite guide and added [Brokered Authentication](https://learn.microsoft.com/en-us/mem/intune/developer/app-sdk-android-phase2#brokered-authentication) as this will be needed for Conditional Access policy to work. There are some additional link in the same doc to verify if brokered authentication is working as expected.

**Comment by bgore** [comment]
We are using Brokered Authentication. Our application requires the MS Company Portal to be present when signing in.  We have verified that the the correct base64 URL encoded signature hash and redirect URI: were created; are in the application's JSON configuration correctly; and has been registered with the application in Azure (per the link above).

For verification, we are using the code sample from above, which attempts to sign-in the user from a ISingleAccountPublicClientApplication.  While enrollment and sign-in fail, as documented above, when I go to Settings -> Accounts, I do see a new account for: 
    <user's email>
    Work account
so I assume that verifies proper broker configuration.

For what it's work, I actually see 2 accounts created.  The other one says:
    Work account
    Work account
I don't know if there was some redundant call or if this is a byproduct of the sign-in failure?

If there is a better/different request to make to verify the broker, let me know; otherwise, what's next?

**Comment by bgore** [comment]
If any help, I've included logcat output from the device, filtered by: "[Intune]" and "companyportal". The former are emitted from our application; the latter from MS CP as it attempts to sign in the user.
[logcat.txt](https://github.com/msintuneappsdk/ms-intune-app-sdk-android/files/15015748/logcat.txt)

**Comment by bgore** [comment]
I guess I'll open an Intune support case, maybe I can get help there.

**Comment by Ddv0623** [comment]
> [@bgore](https://github.com/bgore) Based on the logs, it seems you should be getting the exception MsalIntuneAppProtectionPolicyRequiredException. Can you please check if you have handled it as per the [doc ](https://learn.microsoft.com/en-us/mem/intune/developer/app-sdk-android-phase7#handle-non-compliance-with-msal) and the sample [here ](https://github.com/msintuneappsdk/Taskr-Sample-Intune-Android-App/blob/c11ee86771f7dbc537b2e19e88eaed77ae06b4f0/app/src/main/java/com/microsoft/intune/samples/taskr/MainActivity.java#L206)

Hi,
I am working on make our app support CA that required App Protection.
I do read the document and take a look on the exmaple app.
For iOS app we kind of able to make it work.
but not yet for Android.

So let go thought the login steps:
- install app (user never login)
- user open app try to login
- login be block due to 'App Protection CA' throw MsalIntuneAppProtectionPolicyRequiredException 
- app run remediateCompliance
- MAM run implement MAMServiceAuthenticationCallback of app from registerAuthenticationCallback
- the acquireToken MAMServiceAuthenticationCallback is actuly should be acquireTokenSilent 
- but user are not login so acquireTokenSilent  most likely faild in most case.

From you guys document 
https://learn.microsoft.com/en-us/intune/intune-service/developer/app-sdk-android-phase7#support-app-protection-ca

![Image](https://github.com/user-attachments/assets/943e0860-6476-4e3f-bfef-9a9bf89fa6b1)
Yes, it kind of work 1 time when uninstall Company portal.
But in many case the result is very inconsitent.

Also the exmaple [here ](https://github.com/microsoftconnect/Taskr-Sample-Intune-Android-App/blob/c11ee86771f7dbc537b2e19e88eaed77ae06b4f0/app/src/main/java/com/microsoft/intune/samples/taskr/MainActivity.java#L206) is not implement the remediateCompliance.

please make the implement support App Protection CA more consitent.

**Comment by mukeshk-ms** [comment]
@Ddv0623 the original issue was resolved by installing company portal app and registering the app. See step 4 in the [doc ](https://learn.microsoft.com/en-us/intune/intune-service/protect/app-based-conditional-access-intune#how-app-based-conditional-access-works).

Please check if you have the [right permission](https://learn.microsoft.com/en-us/intune/intune-service/developer/app-sdk-get-started#give-your-app-access-to-the-intune-mobile-app-management-service) for Intune to successfully fetch token silently from Entra Id. Also ensure, you have [granted Admin consent](https://learn.microsoft.com/en-us/entra/identity/enterprise-apps/grant-admin-consent?pivots=portal#grant-admin-consent-in-app-registrations-pane) for the same.
 
Although the sample doesn't implement it completely it has the placeholder with [supporting guidance](https://learn.microsoft.com/en-us/intune/intune-service/developer/app-sdk-android-phase7#mamcompliancemanager) in our doc to implement RemediateCompliance.

Something like following -


    public void onError(@Nullable MsalException exc) {
        if (exc instanceof MsalIntuneAppProtectionPolicyRequiredException) {
    
            final MsalIntuneAppProtectionPolicyRequiredException policyRequiredException =
                (MsalIntuneAppProtectionPolicyRequiredException) ex;
    
            final String upn = policyRequiredException.getAccountUpn();
            final String aadId = policyRequiredException.getAccountUserId();
            final String tenantId = policyRequiredException.getTenantId();
            final String authority = policyRequiredException.getAuthorityURL();
    
            MAMComplianceManager complianceManager = MAMComponents.get(MAMComplianceManager.class);
            complianceManager.remediateCompliance(upn, aadId, tenantId, authority, showUX);
        }
    }

Please create a new issue with additional details if needed to ensure we don't mix different issues. I will close this one.
